apps/dcos_net/src/dcos_net_mesos_state.erl

```diff
@@ -0,0 +1,736 @@
+-module(dcos_net_mesos_state).
+
+%% API
+-export([
+    start_link/0,
+    subscribe/0,
+    is_leader/0,
+    next/1
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2,
+    handle_info/2, terminate/2, code_change/3]).
+
+-export_type([task_id/0, task/0, task_state/0, task_port/0]).
+
+-type task_id() :: binary().
+-type task() :: #{
+    name => binary(),
+    framework => binary() | {id, binary()},
+    agent_ip => inet:ip4_address() | {id, binary()},
+    task_ip => [inet:ip_address()],
+    state => task_state(),
+    ports => [task_port()]
+}.
+-type task_state() :: true | running | {running, boolean()} | false.
+-type task_port() :: #{
+    name => binary(),
+    host_port => inet:port_number(),
+    port => inet:port_number(),
+    protocol => tcp | udp,
+    vip => [binary()]
+}.
+
+-record(state, {
+    pid :: pid(),
+    ref :: reference(),
+    size = undefined :: pos_integer() | undefined,
+    buf = <<>> :: binary(),
+    timeout = 15000 :: timeout(),
+    timeout_ref = make_ref() :: reference(),
+    agents = #{} :: #{binary() => inet:ip4_address()},
+    frameworks = #{} :: #{binary() => binary()},
+    tasks = #{} :: #{task_id() => task()},
+    waiting_tasks = #{} :: #{task_id() => true},
+    subs = undefined :: #{pid() => reference()} | undefined
+}).
+
+-type state() :: #state{}.
+
+-spec(start_link() -> {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec(subscribe() -> {ok, MonRef, Tasks} | {error, atom()}
+    when MonRef :: reference(), Tasks :: #{task_id() => task()}).
+subscribe() ->
+    case whereis(?MODULE) of
+        undefined ->
+            {error, not_found};
+        Pid ->
+            subscribe(Pid)
+    end.
+
+-spec(next(reference()) -> ok).
+next(Ref) ->
+    ?MODULE ! {next, Ref},
+    ok.
+
+-spec(is_leader() -> boolean()).
+is_leader() ->
+    try
+        gen_server:call(?MODULE, is_leader)
+    catch _Class:_Error ->
+        false
+    end.
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    self() ! init,
+    {ok, []}.
+
+handle_call(is_leader, _From, State) ->
+    IsLeader = is_record(State, state),
```

Comment:
It's not immediately clear to me how state being a record translates to IsLeader = true?

Reply:
this gen_server doesn't have state if it isn't connected to mesos master on leader node.

https://github.com/dcos/dcos-net/blob/4d449b72a253204ee8333c17de247470e5105e4e/apps/dcos_net/src/dcos_net_mesos_state.erl#L95-L96
https://github.com/dcos/dcos-net/blob/4d449b72a253204ee8333c17de247470e5105e4e/apps/dcos_net/src/dcos_net_mesos_state.erl#L650-L664